update-engine/src/context.rs

```diff
@@ -24,30 +28,112 @@ use crate::{events::StepProgress, StepSpec};
 /// likely that it is dropped at the same time the future completes.
 #[derive(Debug)]
 pub struct StepContext<S: StepSpec> {
-    progress_sender: mpsc::Sender<StepProgress<S>>,
+    log: slog::Logger,
+    payload_sender: mpsc::Sender<StepContextPayload<S>>,
     token: StepHandleToken<S>,
 }
 
 impl<S: StepSpec> StepContext<S> {
-    pub(crate) fn new(progress_sender: mpsc::Sender<StepProgress<S>>) -> Self {
-        Self { progress_sender, token: StepHandleToken::new() }
+    pub(crate) fn new(
+        log: &slog::Logger,
+        payload_sender: mpsc::Sender<StepContextPayload<S>>,
+    ) -> Self {
+        Self { log: log.clone(), payload_sender, token: StepHandleToken::new() }
     }
 
     /// Sends a progress update to the update engine.
     #[inline]
     pub async fn send_progress(&self, progress: StepProgress<S>) {
-        self.progress_sender
-            .send(progress)
+        self.payload_sender
+            .send(StepContextPayload::Progress(progress))
             .await
             .expect("our code always keeps the receiver open")
     }
 
+    /// Creates a nested execution engine.
+    ///
+    /// An individual step can generate other steps: these steps are treated as
+    /// *nested*, and carry their own progress.
+    pub async fn with_nested_engine<'a, 'this, F, S2>(
+        &'this self,
+        engine_fn: F,
+    ) -> Result<CompletionContext<S2>, S2::Error>
+    where
+        'this: 'a,
+        F: FnOnce(&mut UpdateEngine<'a, S2>) -> Result<(), S2::Error> + Send,
+        S2: StepSpec,
+    {
+        let (sender, mut receiver) = mpsc::channel(128);
+        let mut engine = UpdateEngine::new(&self.log, sender);
+        // Create the engine's steps.
+        (engine_fn)(&mut engine)?;
+
+        // Now run the engine.
+        let engine = engine.execute();
+        tokio::pin!(engine);
+
+        let mut result = None;
+        let mut events_done = false;
+
+        loop {
+            tokio::select! {
+                ret = &mut engine, if result.is_none() => {
```

Comment:
Just double-checking my understanding here: when we enter this `select!` we start polling both `engine` and `receiver.recv()`. If `engine` becomes ready, the future returned by `receiver.recv()` is cancelled (which is fine, `recv()` returns a cancel-safe future) and we handle the result here. If the future returned by `receiver.recv()` becomes ready, `engine` is _not_ cancelled, because we're polling `&mut engine`, right?

Reply:
> If engine becomes ready, the future returned by receiver.recv() is cancelled (which is fine, recv() returns a cancel-safe future) and we handle the result here.

Right, the other part of this is that polling the engine, even if it doesn't complete, drives it forward.

> engine is not cancelled, because we're polling &mut engine, right?

Correct. In general, we can always use the `&mut` pattern to turn any non-cancel-safe future into a cancel-safe one.